ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,104 @@
+
+* We made the decision (at Red Hat) to stop trying to offer a managed service interface for Konflux. The user signup, waitlist, and workspace provisioning interfaces made a lot of sense there.
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. Konflux won't provide its own policy engine for enforcing policies related to
+creating/deleting namespaces. It will instead recommend user to use existing open source projects
+such as [Kyverno](https://kyverno.io/docs/introduction/) and [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/) for this task.
+
+3. We will stop using the term Workspace, and start to use the term Namespace.
+
+4. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
+
+5. Konflux will provide an optional automation for creating a namespace for a user
+when he/she logins for the first time.
+
+6. Konflux won't be opinionated about the mechanism for initializing/maintaining namespaces with
+supporting resources such as ResourceQuota and LimitRange. Konflux will defer
+this responsibility to other tools which are specialized in this task such as 
+(but not limited to) [Kyverno](https://kyverno.io/policies/best-practices/add-ns-quota/add-ns-quota/) and [ArgoCD](https://github.com/konflux-ci/namespace-generator).
+
+7. Konflux will provide a thin [backend service](https://github.com/konflux-ci/workspace-manager) for listing the namespaces where the user has at least view access
+to the Konflux CRDs. This list will be used by the namespace switcher in the UI.
+This is required since the Kubernetes API doesn't let the user to list a subset
+of namespace. The user gets permissions to list all namespace or none.
+
+8. Konflux will provide ClusterRoles that will grant permissions to the Konflux
+and Tekton CRDs. Those will be (aggregated)[https://github.com/konflux-ci/konflux-ci/issues/440] to the built-in Kubernetes roles (`view`, `edit`, `admin`).
+
+9. Public viewer access will be provided by assigning the `view` role (see above) to the
+`system:authenticated` built-in group that contains all the authenticated users.
+
+10. Same as Kubernetes, Konflux won't have a resource for representing a user. Instead,
+it will use external Identity providers.
+
+11. Konflux won't provide a way for creating a ephemeral namespaces. It will differ this
+task to another tool.
+
+12. The `join the waitlist` button will be removed from the Konflux UI.
+
+13. Konflux won't provide a transparent multi-cluster deployment.
+If required, existing open source projects that handle multi-cluster deployments
+should be explored and Konflux should integrate with them.
```

Comment:
Which is the scale at which Konflux aims to operate? Given the Kubernetes tested limits, I'd suggest to consider multi-cluster as something Konflux should be ready to leverage on. I think that limits on namespaces and services (for the largest possible cluster) can be concerning.

The Workspace concept can be an important integration point with multi-cluster management tools (e.g. KubeSaw, OCM, etc) that can facilitate the adoption of Konflux in such complex scenarios.

From https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds

> Many of the thresholds (for cluster scope) are given for the largest possible cluster. For smaller clusters, the limits are proportionally lower.

| Quantity                    | Threshold scope=namespace | Threshold: scope=cluster |
|-----------------------------|---------------------------|--------------------------|
| #Nodes                      | n/a                       | 5000                     |
| #Namespaces                 | n/a                       | 10000                    |
| #Pods                       | 3000                      | 150000                   |
| #Pods per node              | min(110, 10*#cores)       | min(110, 10*#cores)      |
| #Services                   | 5000                      | 10000                    |

For completeness only, here the link to OpenShift limits: https://docs.openshift.com/container-platform/4.17/scalability_and_performance/planning-your-environment-according-to-object-maximums.html#cluster-maximums-major-releases_object-limits

Reply:
This is a good point, but I don't think it's specific to Konflux. It would be nice to find a multicluster solution which is transparent to the application, so it could be used for other applications not just Konflux.

I don't think Konflux should have specific knowledge on any multi-cluster solution or abstraction that may provide it. The reason is that different users may already have their own solution to that (as I said above, it's not a Konflux specific issue).

Reply:
If the are no objections I'll close this thread.